### ISSUE TITLE
"parallel-rustc" isn't working (on my PC at least)

### ISSUE THREAD

**Issue by frederikhors** [rust-lang/rust#117781]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/117781)*

I tried [the instructions](https://blog.rust-lang.org/2023/11/09/parallel-rustc.html#how-to-use-it) using:

- Windows 10 x64
- i7-4790 CPU @ 3.60GHz, Cores: 4, Logical processors: 8
- nightly-x86_64-pc-windows-msvc - rustc 1.75.0-nightly (0f44eb32f 2023-11-09)
- stable-x86_64-pc-windows-msvc - rustc 1.73.0 (cc66ad468 2023-10-03)

I'm using `config.toml` in `.cargo` dir in the project root:

```toml
[build]
rustflags = ["-Z", "threads=8"]
```

### Results

Dev build
- Without, 2m 45s
- With, 2m 47s

Release build
- Without, 11m 37s
- With, 11m 37s (exactly 37s !)

What am I doing wrong?

**Comment by bjorn3** [comment]
(I've moved this issue to the rust repo as the compiler-team repo is meant for organizational purposes and MCP's (major change proposals).)

**Comment by bjorn3** [comment]
Are you using the same nightly both when using `-Zthreads=8` and without `-Zthreads=8`.

**Comment by frederikhors** [comment]
> Are you using the same nightly both when using `-Zthreads=8` and without `-Zthreads=8`.

Yep. With nightly build I'm using `-Zthreads=8`. With stable I'm not using it.

**Comment by bjorn3** [comment]
I mean are you comparing nightly with `-Zthreads=8` against nightly without or nightly with `-Zthreads=8` against stable without?

**Comment by frederikhors** [comment]
Nope. I'm comparing nightly **with** `-Zthreads=8` against stable **without** `-Zthreads=8`.

**Comment by bjorn3** [comment]
Could you try comparing against nightly without `-Zthreads=8` just to check that it works at all?

**Comment by frederikhors** [comment]
Yes I can try right now (I'm also doing a `cargo clean` every time, the `target` directory is recreated from scratch every time) but I think this is pretty obvious that the new parallel frontend isn't working, right?

**Comment by bjorn3** [comment]
Quite some people have reported wins already, so I'm trying to figure out if it is an issue with your local setup somehow ignoring the flag, it not working for your specific project, something windows specific or something else.

**Comment by frederikhors** [comment]
- Nightly, dev, without `-Zthreads=8`: `dev [unoptimized + debuginfo] target(s) in 2m 54s`
- Nightly, dev, with `-Zthreads=8`: `dev [unoptimized + debuginfo] target(s) in 2m 58s`

**Comment by SparrowLii** [comment]
Have you tried using env variable to set RUSTFLAGS? And maybe try other threads like 4 or 16 too

**Comment by lqd** [comment]
And also trying to only build the leaf crate, not the entire project from scratch: to compare results when building a single crate which will have more parallelism opportunities for the frontend.

**Comment by frederikhors** [comment]
> Have you tried using env variable to set RUSTFLAGS? And maybe try other threads like 4 or 16 too

- $env:RUSTFLAGS="-Z threads=8"; cargo build --workspace

    Finished dev [unoptimized + debuginfo] target(s) in 2m 43s

- $env:RUSTFLAGS="-Z threads=4"; cargo build --workspace

    Finished dev [unoptimized + debuginfo] target(s) in 2m 49s

- $env:RUSTFLAGS="-Z threads=16"; cargo build --workspace

    Finished dev [unoptimized + debuginfo] target(s) in 2m 40s

> And also trying to only build the leaf crate, not the entire project from scratch: to compare results when building a single crate which will have more parallelism opportunities for the frontend.

I do not understand what are you asking me to do.

**Comment by R-Goc** [comment]
Did you check that the amount of threads you set is actually being utilized? In my case on Windows it seems that no matter what I set it to it uses 8 threads(all that I have). When I set it to 1 it still uses 8 even when set with $env or the $Env:RUSTFLAGS="".

**Comment by frederikhors** [comment]
How can I check? Anyway, the time doesn't change.

**Comment by R-Goc** [comment]
Just use the task manager

**Comment by HeroicKatora** [comment]
The interesting values are comparing user time vs. wall clock time. For a parallel process the former should vastly exceed the latter as execution time of threads are summed up. On Windows these values can be access via `GetProcessTimes` but I don't recall any command line wrapper to access this conveniently, programmatically anyways, and neither can a quick Google. Maybe the most promising way I've quickly come across would be from [this stackoverflow Q&A](https://stackoverflow.com/questions/27262773/windows-command-line-equivalent-to-time-in-linux/27267238#27267238).

**Comment by frederikhors** [comment]
@HeroicKatora I'm trying this to reduce my compilation times (dev and release). What are you suggesting?

**Comment by HeroicKatora** [comment]
I'm trying to provide a way to measure the answer to:

> Did you check that the amount of threads you set is actually being utilized?

**Comment by Zoxc** [comment]
> I do not understand what are you asking me to do.

Run `cargo check` to build all dependencies. Change your crate (insert a space) and run `cargo check` and observe the time. Do this for both the parallel and non-parallel compiler.

**Comment by frederikhors** [comment]
> Run `cargo check` to build all dependencies. Change your crate (insert a space) and run `cargo check` and observe the time. Do this for both the parallel and non-parallel compiler.

- `cargo clean`

- `cargo check`: 1m 26s

- edited a file in one of the crate of the workspace (added a space in a string)

- `cargo check`: 5s

- `cargo clean`

- added `rustflags = ["-Z", "threads=8"]` under `[build]` in project's root dir/.cargo/config.toml file

- restored the file modified

- `cargo check`: 1m 22s

- edited a file in one of the crate of the workspace (added a space in a string)

- `cargo check`: 5s

I hope this helps, @Zoxc.

**Comment by petrochenkov** [comment]
Triage: technically we could try re-testing this now, but the issue doesn't tell which crate this was tested on.
In general, not all crates get significant benefits from parallel frontend right now, it still has some room for parallelizing more work.
Besides the re-testing, this issue is probably non-actionable and can be closed now.

**Comment by frederikhors** [comment]
@petrochenkov give me an example project to test and I'll report times. Do not close, is still actionable I think. :)

**Comment by mati865** [comment]
@frederikhors you can test on whatever you want. If you need an inspiration, pick any crate from crates.io and check out its code. Just make it's not something totally trivial.

**Comment by petrochenkov** [comment]
> [@petrochenkov](https://github.com/petrochenkov) give me an example project to test and I'll report times. Do not close, is still actionable I think. :)

"Please find a crate on which the parallel frontend doesn't work well" is not a good issue to keep open.
Before the stabilization we'll compare times "1 vs 2 vs 4 vs 8 vs 16 threads" on the standard rustc-perf benchmarks to prove that the parallel frontend is generally useful.
